test/integration/protocol_integration_test.cc

```diff
@@ -4921,4 +4921,64 @@ TEST_P(DownstreamProtocolIntegrationTest, UnknownPseudoHeader) {
   }
 }
 
+// This test performs DNS refreshing very fast also keep changing DNS results.
+// It is trying to reproduce a R/W race for logical_dns_cluster.
+class LogicalDnsReadWriteRaceTest : public ProtocolIntegrationTest {
+public:
+  LogicalDnsReadWriteRaceTest() : registered_dns_factory_(dns_resolver_factory_) {}
+
+  NiceMock<Network::MockDnsResolverFactory> dns_resolver_factory_;
+  Registry::InjectFactory<Network::DnsResolverFactory> registered_dns_factory_;
+};
+
+INSTANTIATE_TEST_SUITE_P(
+    Protocols, LogicalDnsReadWriteRaceTest,
+    testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParamsWithoutHTTP3()),
+    HttpProtocolIntegrationTest::protocolTestParamsToString);
+
+TEST_P(LogicalDnsReadWriteRaceTest, FastChangingDnsResult) {
+  if (version_ == Network::Address::IpVersion::v4) {
+    return;
+  }
+  std::shared_ptr<Network::MockDnsResolver> dns_resolver(new Network::MockDnsResolver());
+  EXPECT_CALL(dns_resolver_factory_, createDnsResolver(_, _, _))
+      .WillRepeatedly(testing::Return(dns_resolver));
+  EXPECT_CALL(*dns_resolver, resolve(_, _, _))
+      .WillRepeatedly(
+          Invoke([&](const std::string&, Network::DnsLookupFamily,
+                     Network::DnsResolver::ResolveCb dns_callback) -> Network::ActiveDnsQuery* {
+            // Keep changing the DNS response address list.
+            static int i = 1;
+            dns_callback(Network::DnsResolver::ResolutionStatus::Success,
+                         TestUtility::makeDnsResponse({"::1", ("127.0.0." + std::to_string(i)),
```

Comment:
`absl::StrCat("127.0.0.", i)`

Reply:
done in https://github.com/envoyproxy/envoy/pull/32830